File: model/agent.go
package model

type AgentSessionStatus string
type AgentMessageRole string
type AgentRunStatus string
type AgentStepType string
type AgentStepStatus string
type AgentEventType string
type AgentMemoryKind string
type AgentMemoryStatus string

const (
	AgentSessionStatusActive   AgentSessionStatus = "active"
	AgentSessionStatusArchived AgentSessionStatus = "archived"

	AgentMessageRoleUser      AgentMessageRole = "user"
	AgentMessageRoleAssistant AgentMessageRole = "assistant"

	AgentRunStatusRunning             AgentRunStatus = "running"
	AgentRunStatusWaitingConfirmation AgentRunStatus = "waiting_confirmation"
	AgentRunStatusWaitingTool         AgentRunStatus = "waiting_tool"
	AgentRunStatusCompleted           AgentRunStatus = "completed"
	AgentRunStatusFailed      AgentRunStatus = "failed"
	AgentRunStatusCancelled   AgentRunStatus = "cancelled"

	AgentStepTypeCompletion AgentStepType = "completion"
	AgentStepTypeTool       AgentStepType = "tool"

	AgentStepStatusRunning   AgentStepStatus = "running"
	AgentStepStatusCompleted AgentStepStatus = "completed"
	AgentStepStatusFailed    AgentStepStatus = "failed"
	AgentStepStatusCancelled AgentStepStatus = "cancelled"

	AgentEventRunStarted           AgentEventType = "run.started"
	AgentEventPlanCreated          AgentEventType = "plan.created"
	AgentEventMessageDelta         AgentEventType = "message.delta"
	AgentEventToolConfirmationRequired AgentEventType = "tool.confirmation_required"
	AgentEventToolCall             AgentEventType = "tool.call"
	AgentEventToolCompleted AgentEventType = "tool.completed"
	AgentEventToolReverted  AgentEventType = "tool.reverted"
	AgentEventRunCompleted  AgentEventType = "run.completed"
	AgentEventRunFailed     AgentEventType = "run.failed"
	AgentEventRunCancelled  AgentEventType = "run.cancelled"

	AgentMemoryKindPreference AgentMemoryKind = "preference"
	AgentMemoryKindFact       AgentMemoryKind = "fact"
	AgentMemoryKindConstraint AgentMemoryKind = "constraint"
	AgentMemoryKindExperience AgentMemoryKind = "experience"

	AgentMemoryStatusActive   AgentMemoryStatus = "active"
	AgentMemoryStatusForgotten AgentMemoryStatus = "forgotten"
)

type AgentSession struct {
	ID             string             `json:"id" gorm:"primaryKey"`
	OrganizationID string             `json:"organizationId" gorm:"index:idx_agent_session_owner,priority:1"`
	UserID         string             `json:"userId" gorm:"index:idx_agent_session_owner,priority:2"`
	ProjectID      string             `json:"projectId" gorm:"index"`
	Profile        string             `json:"profile" gorm:"type:text"`
	Title          string             `json:"title"`
	Status         AgentSessionStatus `json:"status" gorm:"index"`
	CreatedAt      string             `json:"createdAt" gorm:"index"`
	UpdatedAt      string             `json:"updatedAt"`
}

type AgentMessage struct {
	ID             string           `json:"id" gorm:"primaryKey"`
	OrganizationID string           `json:"organizationId" gorm:"index:idx_agent_message_owner,priority:1"`
	UserID         string           `json:"userId" gorm:"index:idx_agent_message_owner,priority:2"`
	SessionID      string           `json:"sessionId" gorm:"uniqueIndex:idx_agent_message_sequence,priority:1;index"`
	Role           AgentMessageRole `json:"role" gorm:"index"`
	Content        string           `json:"content" gorm:"type:text"`
	Sequence       int64            `json:"sequence" gorm:"uniqueIndex:idx_agent_message_sequence,priority:2"`
	CreatedAt      string           `json:"createdAt" gorm:"index"`
}

type AgentRun struct {
	ID             string         `json:"id" gorm:"primaryKey"`
	OrganizationID string         `json:"organizationId" gorm:"index:idx_agent_run_owner,priority:1"`
	UserID         string         `json:"userId" gorm:"index:idx_agent_run_owner,priority:2"`
	SessionID      string         `json:"sessionId" gorm:"index"`
	MessageID      string         `json:"-" gorm:"index"`
	Model          string         `json:"model" gorm:"index"`
	Context        string         `json:"-" gorm:"type:text"`
	Status         AgentRunStatus `json:"status" gorm:"index"`
	Error          string         `json:"error,omitempty" gorm:"type:text"`
	StartedAt      string         `json:"startedAt" gorm:"index"`
	CompletedAt    string         `json:"completedAt,omitempty"`
	CreatedAt      string         `json:"createdAt" gorm:"index"`
	UpdatedAt      string         `json:"updatedAt"`
}

type AgentStep struct {
	ID             string          `json:"id" gorm:"primaryKey"`
	OrganizationID string          `json:"organizationId" gorm:"index:idx_agent_step_owner,priority:1"`
	UserID         string          `json:"userId" gorm:"index:idx_agent_step_owner,priority:2"`
	RunID          string          `json:"runId" gorm:"index;index:idx_agent_step_tool_call,priority:1"`
	Type           AgentStepType   `json:"type" gorm:"index"`
	Status         AgentStepStatus `json:"status" gorm:"index"`
	ToolCallID     string          `json:"toolCallId" gorm:"index:idx_agent_step_tool_call,priority:2"`
	ToolName       string          `json:"toolName" gorm:"index"`
	Confirmation   string          `json:"confirmation,omitempty" gorm:"index"`
	ExecutionToken string          `json:"-" gorm:"index"`
	ExecutionAt    string          `json:"-" gorm:"index"`
	Input          string          `json:"-" gorm:"type:text"`
	Output         string          `json:"-" gorm:"type:text"`
	Error          string          `json:"error,omitempty" gorm:"type:text"`
	StartedAt      string          `json:"startedAt"`
	CompletedAt    string          `json:"completedAt,omitempty"`
	CreatedAt      string          `json:"createdAt"`
	UpdatedAt      string          `json:"updatedAt"`
}

type AgentEvent struct {
	ID             string         `json:"id" gorm:"primaryKey"`
	OrganizationID string         `json:"organizationId" gorm:"index:idx_agent_event_owner,priority:1"`
	UserID         string         `json:"userId" gorm:"index:idx_agent_event_owner,priority:2"`
	RunID          string         `json:"runId" gorm:"uniqueIndex:idx_agent_event_sequence,priority:1;index"`
	Sequence       int64          `json:"sequence" gorm:"uniqueIndex:idx_agent_event_sequence,priority:2"`
	Type           AgentEventType `json:"type" gorm:"index"`
	Payload        string         `json:"-" gorm:"type:text"`
	CreatedAt      string         `json:"createdAt" gorm:"index"`
}

// AgentMemory is a user-controlled, server-side fact that can be reused by later Runs.
// Content is intentionally plain text and never contains model credentials or canvas media.
type AgentMemory struct {
	ID             string            `json:"id" gorm:"primaryKey"`
	OrganizationID string            `json:"organizationId" gorm:"index:idx_agent_memory_scope,priority:1;uniqueIndex:idx_agent_memory_identity,priority:1"`
	UserID         string            `json:"userId" gorm:"index:idx_agent_memory_scope,priority:2;uniqueIndex:idx_agent_memory_identity,priority:2"`
	ProjectID      string            `json:"projectId" gorm:"index:idx_agent_memory_scope,priority:3;uniqueIndex:idx_agent_memory_identity,priority:3"`
	Kind           AgentMemoryKind   `json:"kind" gorm:"index"`
	Key            string            `json:"key" gorm:"index:idx_agent_memory_key,priority:1;uniqueIndex:idx_agent_memory_identity,priority:4"`
	Content        string            `json:"content" gorm:"type:text"`
	SourceRunID    string            `json:"sourceRunId,omitempty" gorm:"index"`
	SourceMessageID string           `json:"sourceMessageId,omitempty" gorm:"index"`
	Confidence     float64           `json:"confidence"`
	Status         AgentMemoryStatus `json:"status" gorm:"index"`
	ExpiresAt      string            `json:"expiresAt,omitempty" gorm:"index"`
	ForgottenAt    string            `json:"forgottenAt,omitempty"`
	CreatedAt      string            `json:"createdAt" gorm:"index"`
	UpdatedAt      string            `json:"updatedAt"`
}

func (run AgentRun) Terminal() bool {
	return run.Status == AgentRunStatusCompleted || run.Status == AgentRunStatusFailed || run.Status == AgentRunStatusCancelled
}

File: web/src/services/api/agent.ts
import { apiGet, apiPost, organizationHeaders } from "./request";

export type AgentSession = {
    id: string;
    projectId: string;
    profile: string;
    title: string;
    status: "active" | "archived";
    createdAt: string;
    updatedAt: string;
};

export type AgentMessage = {
    id: string;
    sessionId: string;
    role: "user" | "assistant";
    content: string;
    sequence: number;
    createdAt: string;
};

export type AgentRun = {
    id: string;
    sessionId: string;
    model: string;
    status: "running" | "waiting_tool" | "waiting_confirmation" | "completed" | "failed" | "cancelled";
    error?: string;
    startedAt: string;
    completedAt?: string;
    maxToolCalls: number;
    maxMediaCalls: number;
    maxDurationSec: number;
    maxCredits: number;
    budgetReason?: "tool_calls" | "media_calls" | "duration" | "credits";
    streamReconnects: number;
    toolLeaseTakeovers: number;
};

export type AgentRunBudget = { maxToolCalls: number; maxMediaCalls: number; maxDurationSec: number; maxCredits: number };

export type AgentCanvasNode = {
    id: string;
    type: string;
    title: string;
    x: number;
    y: number;
    width: number;
    height: number;
    content?: string;
    prompt?: string;
    references?: string[];
    storageKey?: string;
};

export type AgentCanvasConnection = {
    from: string;
    to: string;
};

export type AgentCanvasContext = {
    autonomy: "cautious" | "standard" | "autonomous";
    selectedNodeIds: string[];
    focusNodeIds: string[];
    nodes: AgentCanvasNode[];
    connections: AgentCanvasConnection[];
};

export type AgentToolArguments =
    | { summary: string; steps: string[] }
    | { prompt: string; count: number; referenceNodeIds?: string[] }
    | { nodeId: string; prompt: string; count: number }
    | { nodeIds: string[]; criteria: string }
    | { prompt: string; duration: number; imageNodeId?: string }
    | { nodeId: string; criteria: string }
    | { nodeIds: string[]; mode: "horizontal" | "vertical" | "grid"; gap: number }
    | { text: string; placement: "center" | "right_of_selection"; sourceNodeIds?: string[] }
    | { nodeIds: string[] }
    | { nodeId: string; text: string }
    | { question: string; options: string[] }
    | { kind: "preference" | "fact" | "constraint" | "experience"; key: string; content: string; scope: "project" | "user"; confidence: number; expiresInDays: number }
    | { key: string; scope: "project" | "user" };

export type AgentToolName =
    "canvas.plan" | "image.generate" | "image.edit" | "image.inspect" | "video.generate" | "video.inspect" | "canvas.arrange" | "canvas.add_text" | "canvas.delete" | "canvas.update_text" | "agent.ask_user" | "agent.remember" | "agent.forget";

export type AgentToolInspection = {
    status: "passed" | "needs_revision" | "unavailable";
    summary: string;
    issues: string[];
    revisedPrompt?: string;
};

export type AgentEvent = {
    id: string;
    runId: string;
    sequence: number;
    type: "run.started" | "plan.created" | "message.delta" | "tool.confirmation_required" | "tool.call" | "tool.completed" | "tool.reverted" | "run.completed" | "run.failed" | "run.cancelled";
    data: {
        content?: string;
        error?: string;
        status?: "success" | "failed" | "approved" | "rejected";
        reason?: "tool_reverted";
        callId?: string;
        name?: AgentToolName;
        arguments?: AgentToolArguments;
        output?: { answer?: string; inspection?: AgentToolInspection; memory?: { id?: string; key: string; status: "active" | "forgotten"; scope: "project" | "user" } };
        meta?: { needsClaim?: boolean; retryOf?: string };
    };
    createdAt: string;
};

export function createAgentSession(projectId: string, title = "新对话", sessionId?: string) {
    return apiPost<AgentSession>("/api/v1/agent/sessions", { sessionId, projectId, title, profile: {} });
}

export function submitAgentMessage(sessionId: string, runId: string, content: string, model: string, canvasContext: AgentCanvasContext, canvasSnapshot: unknown, budget: AgentRunBudget) {
    return apiPost<{ message: AgentMessage; run: AgentRun }>(`/api/v1/agent/sessions/${sessionId}/messages`, { runId, content, model, canvasContext, canvasSnapshot, budget });
}

export type AgentToolResult =
    | { callId: string; status: "success"; plan: { summary: string; steps: string[] }; inspection?: never; images?: never; video?: never; nodeIds?: never; positions?: never; nodeId?: never; text?: never; placement?: never; memory?: never; error?: never }
    | {
          callId: string;
          status: "success";
          images: { nodeId: string; storageKey: string }[];
          plan?: never;
          inspection?: never;
          video?: never;
          nodeIds?: never;
          positions?: never;
          nodeId?: never;
          text?: never;
          placement?: never;
          memory?: never;
          error?: never;
      }
    | { callId: string; status: "success"; inspection: AgentToolInspection; plan?: never; images?: never; video?: never; nodeIds?: never; positions?: never; nodeId?: never; text?: never; placement?: never; memory?: never; error?: never }
    | {
          callId: string;
          status: "success";
          video: { nodeId: string; storageKey: string };
          plan?: never;
          inspection?: never;
          images?: never;
          nodeIds?: never;
          positions?: never;
          nodeId?: never;
          text?: never;
          placement?: never;
          memory?: never;
          error?: never;
      }
    | {
          callId: string;
          status: "success";
          nodeIds: string[];
          positions: { nodeId: string; x: number; y: number }[];
          plan?: never;
          inspection?: never;
          images?: never;
          video?: never;
          nodeId?: never;
          text?: never;
          placement?: never;
          memory?: never;
          error?: never;
      }
    | { callId: string; status: "success"; nodeId: string; placement: "center" | "right_of_selection"; plan?: never; inspection?: never; images?: never; video?: never; nodeIds?: never; positions?: never; text?: never; memory?: never; error?: never }
    | { callId: string; status: "success"; nodeIds: string[]; plan?: never; inspection?: never; images?: never; video?: never; positions?: never; nodeId?: never; text?: never; placement?: never; memory?: never; error?: never }
    | { callId: string; status: "success"; nodeId: string; text: string; plan?: never; inspection?: never; images?: never; video?: never; nodeIds?: never; positions?: never; placement?: never; memory?: never; error?: never }
    | {
          callId: string;
          status: "success";
          memory: { id?: string; key: string; status: "active" | "forgotten"; scope: "project" | "user" };
          plan?: never;
          inspection?: never;
          images?: never;
          video?: never;
          nodeIds?: never;
          positions?: never;
          nodeId?: never;
          text?: never;
          placement?: never;
          error?: never;
      }
    | { callId: string; status: "failed"; error: string; plan?: never; inspection?: never; images?: never; video?: never; nodeIds?: never; positions?: never; nodeId?: never; text?: never; placement?: never; memory?: never };

export function submitAgentToolResult(runId: string, executionToken: string, result: AgentToolResult) {
    return apiPost<AgentRun>(`/api/v1/agent/runs/${runId}/tool-results`, { ...result, executionToken });
}

export function claimAgentToolExecution(runId: string, callId: string, token: string) {
    return apiPost<{ status: "claimed" }>(`/api/v1/agent/runs/${runId}/tool-claims`, { callId, token });
}

export function getAgentRun(runId: string) {
    return apiGet<AgentRun>(`/api/v1/agent/runs/${runId}`);
}

export type AgentStepDiagnostic = {
    callId?: string;
    type: "completion" | "tool";
    toolName?: AgentToolName;
    status: "running" | "completed" | "failed" | "cancelled";
    confirmation?: "approved" | "rejected";
    error?: string;
    startedAt: string;
    completedAt?: string;
    durationMs: number;
    retryable: boolean;
    revertible: boolean;
    reverted: boolean;
};

export type AgentRunDiagnostics = Pick<AgentRun, "id" | "status" | "model" | "error" | "startedAt" | "completedAt"> & {
    durationMs: number;
    canRevert: boolean;
    budget: AgentRunBudget;
    usage: { toolCalls: number; mediaCalls: number; durationSec: number; streamReconnects: number; credits: number; toolLeaseTakeovers: number };
    budgetReason?: "tool_calls" | "media_calls" | "duration" | "credits";
    plan: { id: string; revision: number; position: number; title: string; completionCriteria: string; dependsOnPosition?: number; status: "pending" | "running" | "completed" | "failed" | "skipped"; toolName?: AgentToolName; reason?: string }[];
    steps: AgentStepDiagnostic[];
};

export function getAgentRunDiagnostics(runId: string) {
    return apiGet<AgentRunDiagnostics>(`/api/v1/agent/runs/${runId}/diagnostics`);
}

export function retryAgentStep(runId: string, callId: string) {
    return apiPost<{ run: AgentRun; callId: string; sourceCallId: string }>(`/api/v1/agent/runs/${runId}/steps/${encodeURIComponent(callId)}/retry`);
}

export function cancelAgentRun(runId: string) {
    return apiPost<AgentRun>(`/api/v1/agent/runs/${runId}/cancel`);
}

export function revertAgentTool(runId: string, callId: string) {
    return apiPost<AgentRun>(`/api/v1/agent/runs/${runId}/tool-reverts`, { callId });
}

export function revertAgentRun(runId: string) {
    return apiPost<{ run: AgentRun; callIds: string[]; snapshot: { nodes: unknown[]; connections: unknown[] }; snapshotChecksum: string }>(`/api/v1/agent/runs/${runId}/revert`);
}

export type AgentFeedbackSignal = "accepted" | "helpful" | "unhelpful" | "deleted" | "corrected";

export function submitAgentFeedback(runId: string, signal: AgentFeedbackSignal, note?: string) {
    return apiPost<{ feedback: { id: string; runId: string; signal: AgentFeedbackSignal; note?: string }; adjustedMemories: number }>(`/api/v1/agent/runs/${runId}/feedback`, { signal, note });
}

export type AgentMetrics = {
    hours: number;
    runs: number;
    completedRuns: number;
    failedRuns: number;
    cancelledRuns: number;
    budgetStoppedRuns: number;
    toolCalls: number;
    failedToolCalls: number;
    mediaCalls: number;
    streamReconnects: number;
    toolLeaseTakeovers: number;
    feedback: Record<string, number>;
    averageDurationMs: number;
    averageToolCalls: number;
    toolFailureRate: number;
    completionRate: number;
    credits: number;
    alerts: string[];
};

export function getAgentMetrics(hours = 24) {
    return apiGet<AgentMetrics>(`/api/v1/agent/metrics?hours=${hours}`);
}

export function getAgentToolResultReceipt(runId: string, callId: string) {
    return apiGet<{ status: "pending" | "completed" | "rejected" | "failed" | "cancelled"; result?: AgentToolResult }>(`/api/v1/agent/runs/${runId}/tool-results/${encodeURIComponent(callId)}`);
}

export function confirmAgentTool(runId: string, callId: string, decision: "approved" | "rejected", answer?: string) {
    return apiPost<AgentRun>(`/api/v1/agent/runs/${runId}/confirmation`, { callId, decision, answer });
}

export async function streamAgentRun(runId: string, onEvent: (event: AgentEvent) => void | Promise<void>, signal?: AbortSignal, after = 0, onProcessed?: (sequence: number) => void) {
    const response = await fetch(`/api/v1/agent/runs/${runId}/events${after > 0 ? `?after=${after}` : ""}`, {
        credentials: "include",
        headers: organizationHeaders(),
        signal,
    });
    if (!response.ok || !response.body) throw new Error("助手事件流连接失败");

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        const blocks = buffer.split("\n\n");
        buffer = blocks.pop() || "";
        for (const block of blocks) {
            const data = block
                .split("\n")
                .find((line) => line.startsWith("data: "))
                ?.slice(6);
            if (data) {
                const event = JSON.parse(data) as AgentEvent;
                await onEvent(event);
                onProcessed?.(event.sequence);
            }
        }
    }
}
